### PR TITLE
[Merged by Bors] - fix: session ID standarization [bugfix] (VF-4058)

### PR DESCRIPTION
**Pull request by DecathectZero**
**Fixes or implements VF-4058**

### Brief description. What is this change?
We need a unique userID for every session, so we can track where the user is on the diagram and where to go to next.
This is automatically generated by dialogflow and provided there. The `req.session` variable can vary drastically by DF channel:

WEB DEMO:
`projects/brand-new-test-a7b4b/agent/sessions/webdemo-0eacbe49-1cf1-8089-201a-916bbf37e738`

DF MESSENGER:
`projects/brand-new-test-a7b4b/agent/sessions/dfMessenger-44661107`

"Try It now sidebar":
`projects/brand-new-test-a7b4b/agent/sessions/e15e10ed-fcfc-204b-9f3a-f7d47b272388`

TWILIO:
`projects/brand-new-test-a7b4b/agent/sessions/065j7JqgM2_Q9iMDpSbrSto3A`

GATEWAY:
`projects/brand-new-test-a7b4b/agent/sessions/11973HDWr0zSQOm6Wx6OJUVww`
`projects/brand-new-test-a7b4b/locations/global/agent/sessions/11973HDWr0zSQOm6Wx6OJUVww`

The real bug is that we would use the entire `req.session` value as the index for the state, when we only need to use the part after `/session/{HERE}`

I've made it so we extract out just that part, i.e. `dfMessenger-44661107` from `projects/brand-new-test-a7b4b/agent/sessions/dfMessenger-44661107` and use that as our actual sessionID

I've also added a temporary migration system for existing prod sessions. After a week we can remove it. This would be equivalent to clearing all sessions older than a week?

### CODE REVIEW

File: lib/services/dialogflow/index.ts [hunk 1]
```diff
@@ -28,22 +30,38 @@ class DialogflowManager extends AbstractManager<{
...
+
+    /**
+     * migrate session IDs
+     * safely remove after 07/20/2022
+     */
+    const oldSessionState = await state?.getFromDb<State>(req.session);
+    if (oldSessionState?.stack) {
+      await state.saveToDb(userID, oldSessionState);
+      await state.deleteFromDb(req.session);
+    }
```
Comment (DecathectZero) [inline lib/services/dialogflow/index.ts:57]: this is not pretty and ideally we can just remove it entirely after we're ok with dropping sessions.
It's better to not run a massive migration :/

File: lib/services/dialogflow/index.ts [hunk 2]
```diff
@@ -28,22 +30,38 @@ class DialogflowManager extends AbstractManager<{
   responseES: ResponseES;
   slotFillingES: SlotFillingES;
 }> {
+  static extractSessionID(session: string) {
+    const path = session.split('/');
+    const sessionIndex = path.indexOf('session');
+    return (sessionIndex > -1 && path[sessionIndex + 1]) || path.splice(-1)[0];
+  }
```
Comment (jonahsnider) [inline lib/services/dialogflow/index.ts:37]: i would use regular expressions here instead of doing it manually with indexes

**Review by jonahsnider — APPROVED** [review]

**Comment by zhihil** [comment]
Perhaps we should be mapping to `projects/:projectID/sessions/:sessionID` instead. I wonder if the `sessionID` isn't sufficiently random and we'll need the `:projectID` to avoid collisions.

The other tokens including `"global"` and `"location"` are common to all of the examples you've given so far, so those don't give any useful information for distinguishing the sessions.

File: lib/services/dialogflow/index.ts (hunk 2 above)
Comment (DecathectZero) [reply lib/services/dialogflow/index.ts:37]: we also want to just default to the last item in the chain if it's not found `path.splice(-1)[0];`

**Pull request closed without merging**